### PR TITLE
CONSF-45: InfoboxDataModel

### PR DESCRIPTION
**Pull request by kenkouot**
## Proposed API

The following code example illustrates the proposed API for working with the structural nature of the infobox data.

``` es6
// Instantiate a new instance of the infobox
var infobox = new InfoboxTemplateBuilder({});

// alias the infobox data model for easier handling
var infoData = infobox.data;

// Create a new field
var title = infoData.newElement('Field', {
    defaultValue: 'Dang',
    label: 'Who let the dogs out?',
    boundVariableName: 'foo'
});

// And add it to the infobox data model instance
infoData.add(title);

// Or create a new group, with that field
var group = infoData.newElement('Group', {
    items: [title]
});

// Add more fields to that group
group.add(infoData.newElement('Image', {
    defaultValue: 'dang.png',
    label: 'Avatar',
    boundVariableName: 'bar'
}));

// and finally add that group to the infobox data instance
infoData.add(group);
```

And here is an example of an instance of `InfoboxData` serialized to JSON.

``` json
{
    "validators": {},
    "items": [
        {
            "validators": {},
            "_nodeType": "data",
            "boundVariableName": "foo",
            "defaultValue": "Dang",
            "label": "Who let the dogs out?",
            "stringTemplate": null
        },
        {
            "validators": {},
            "items": [
                {
                    "validators": {},
                    "_nodeType": "data",
                    "boundVariableName": "foo",
                    "defaultValue": "Dang",
                    "label": "Who let the dogs out?",
                    "stringTemplate": null
                },
                {
                    "validators": {},
                    "_nodeType": "image",
                    "boundVariableName": "bar",
                    "defaultValue": "dang.png",
                    "alt": null,
                    "caption": null,
                    "label": "Avatar"
                }
            ],
            "_nodeType": "group",
            "layout": null,
            "show": null
        }
    ],
    "theme": null,
    "themeVarName": null,
    "title": null,
    "layout": null
}
```

### CODE REVIEW

File: src/core.js [hunk 1]
```diff
@@ -24,7 +38,8 @@ class Core extends Model {
 
 		} else {
 
-			this.data = null; // new InfoboxData(...);
+			// If no 'from' is not defend, we instantiate a fresh infobox
```
Comment (lizlux) [inline src/core.js]: typo in comment

File: src/core.js [hunk 2]
```diff
@@ -1,7 +1,17 @@
...
-import {Model} from './model';
-import {serialize, deserialize} from './serializers/xml';
+import {InfoboxData} from './models/infobox-data';
+import {Model} from './models/base';
 import {persist} from './adapters/mediawiki';
+import {serialize, deserialize} from './serializers/xml';
+
+const defaultProps = {
+	// Options to be passed to InfoboxData constructor
+	infoboxOptions: null,
```
Comment (lizlux) [inline src/core.js]: How about `dataOptions`? To match `themeOptions`

File: src/core.js (hunk 1 above)
Comment (lizlux) [inline src/core.js]: actually two typos. I think you mean:  `If 'from' is not defined, we instantiate a fresh infobox`

File: test.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
+// Instantiate a new instance of the infobox
```
Comment (lizlux) [inline test.js:1]: Is the intention to leave this in the repo or is this just a convenience for development? Would be good to turn it into an actual unit test.

File: src/types/field.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	constructor(properties = {}) {
+		super();
+		this._model = 'field';
+		Object.assign(this, AttributeMixin, defaultProperties, properties);
+	}
+
+	set(propName, value) {
+
+		if (typeMap[propName]) {
+			const isValid = typeMap[propName].validator();
```
Comment (lizlux) [inline src/types/field.js]: Based on the current code, this should be just `typeMap[propName]()`

File: src/types/field.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+			if (!isValid) {
+				throw new TypeError(`${propName} should be of ${typeMap[propName].type}`);
```
Comment (lizlux) [inline src/types/field.js]: And based on current code, `typeMap[propName].type` doesn't exist.

File: src/types/mixins.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
+'use strict';
+import {isString} from '../validators';
+
+/**
+ * AttributeMixin
+ * @description This is a mixin for adding attribute support to a node
+ * Attributes are simple key:value pairs meant to store any client specific implementation details for a node
+ * @mixin
+ */
+export const AttributeMixin = Object.create(null, {
```
Comment (lizlux) [inline src/types/mixins.js]: Let's chat about the need for this mixin when you're in the office to make sure it's not overkill.

Comment (kenkouot) [inline src/types/mixins.js]: I'm going to remove it actually

File: test.js (hunk 3 above)
Comment (kenkouot) [inline test.js:1]: It's a convenience for now, tests are being handled separately.

File: src/core.js [hunk 7]
```diff
@@ -9,6 +19,10 @@ class Core extends Model {
 
 		super();
 
+		// extend the properties
+		params = Object.assign(defaultProps, params);
+
+		// The for
```
Comment (lizlux) [inline src/core.js]: ?

File: src/models/collection.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+		this.setItems(itemsCopy);
+		return item;
+	}
+
+	remove(index) {
+		const itemsCopy = copyArray(this.items);
+		let removed;
+
+		if (index) {
```
Comment (lizlux) [inline src/models/collection.js:34]: let's check if `index` is a number

File: src/utils.js [hunk 9]
```diff
@@ -35,3 +35,20 @@ export function deepSet(str, val, context = this) {
...
+ */
+export function copyArray(arr = []) {
+	return arr.slice(0);
+}
+
+/*
+ * Given an array and two indices, swap the elements in place.
+ */
+export function swapArrayElements(arr, idx1, idx2) {
+	const tmp = arr[idx2];
```
Comment (lizlux) [inline src/utils.js:57]: might be worth checking if `idx1` and `idx2` are numbers here too.

**Comment by lizlux** [comment]
LGTM

**Pull request merged**